### DISCUSSION TITLE
simulation/turbine.py - power function return

### DISCUSSION THREAD

**Discussion by ryan-redman** [NatLabRockies/floris#611] [Q&A] [answered]
Lines 141 -150 in Floris V3.2.1

The power function returns p * ref_density_cp_ct. 
p is the turbine power curve interpolated to the effective yaw velocities. Why would the power curve need to be multiplied by the reference air density?

I'm also confused by the comments about thrust, when the calculation only involves power.

```
# Loop over each turbine type given to get thrust coefficient for all turbines
    p = np.zeros(np.shape(yaw_effective_velocity))
    power_interp = dict(power_interp)
    turb_types = np.unique(turbine_type_map)
    for turb_type in turb_types:
        # Using a masked array, apply the thrust coefficient for all turbines of the current
        # type to the main thrust coefficient array
        p += power_interp[turb_type](yaw_effective_velocity) * np.array(turbine_type_map == turb_type)

    return p * ref_density_cp_ct
```

**Comment by Bartdoekemeijer** [comment, marked as answer]
Hi @ryan-redman. This is implemented to accommodate the situation where the air density at which the turbine cp and ct tables are specified differs from the air density that is assumed in the FLORIS simulation. See #398.

**Reply by RyanDunn729** [reply]
Hi @Bartdoekemeijer , I encountered the same question when looking through this. It appears that the adjustment for different air densities is handled in the "yaw_effective_velocity" variable, see lines 141-145
```python
    yaw_effective_velocity = (
        (air_density/ref_density_cp_ct)**(1/3)
        * average_velocity(velocities)
        * cosd(yaw_angle) ** pW
    )
```

Can you give more explanation as to why (p * ref_density_cp_ct) is correct? It seems like it is multiplying Cp*[kg/m3]. The only explanation I see is that the data within the .yaml files are assumed to be normalized by 1/ref_density_cp_ct. (the same does not occur for the Ct( ) function)

**Reply by paulf81** [reply]
I think this is related to the same points raised in discussion #729, Cp->power leads to a high sensitivity to small numerical variations (1~2kW at 5000kW), that warranted the move to interpolating on power instead of Cp.  Thurst is not used itself, and such small changes to Ct, at the time, didn't seem to necessitate the same move, so Ct is kept more simple.

If you check the definition of the power function interpolated over defined in the init function:

```
        inner_power = (
            0.5 * self.rotor_area
            * self.fCp_interp(wind_speeds)
            * self.generator_efficiency
            * wind_speeds ** 3
        )
```

You see it excludes air density, so that this can be modified, for example by the reinitialize function, in code.   This function:

```
 yaw_effective_velocity = (
        (air_density/ref_density_cp_ct)**(1/3)
        * average_velocity(velocities)
        * cosd(yaw_angle) ** pW
    )
```

Has the intention that if the air density specified is different than the one which defines the input cp/ct, then the effective_velocity is adjusted (incorporating the change this way better handles region 3 versus simply multiplying by air_density, which would cause a drop in power in above-rated conditions in response to a change in air density when in reality the control system would react to keep things rated)

Then with this correction, the original ref_density_cp_ct can be multiplied at the last step.
